### PR TITLE
Check for Feature Flag.

### PR DESCRIPTION
**Pull request by AdamSpeight2008**
Modifed check feature availability to check for the presence of Feature Flag before declaring it being not found.

### CODE REVIEW

**Comment by AdamSpeight2008** [comment]
'D:\j\workspace\windows_debug---09a07fba\src\Compilers\VisualBasic\Portable\Syntax\SyntaxFactory.vb(1114,60): error BC30512: Option Strict On disallows implicit conversions from 'LanguageVersion' to 'Feature'. [D:\j\workspace\windows_debug---09a07fba\src\Compilers\VisualBasic\Portable\BasicCodeAnalysis.vbproj]`

LIne 1114 ? of  a 537 line file.

**Comment by AdamSpeight2008** [comment]
retest this please

**Comment by AdamSpeight2008** [comment]
@dotnet-bot retest this please

**Comment by AdamSpeight2008** [comment]
retest this please

**Comment by AdamSpeight2008** [comment]
@dotnet-bot retest this please

**Comment by AdamSpeight2008** [comment]
@dotnet-bot test this please

**Comment by AdamSpeight2008** [comment]
This builds and passes all tests locally.
roslynbuild

**Comment by AdamSpeight2008** [comment]
@dotnet-bot please retest windows_eta_open_prtest

**Comment by AdamSpeight2008** [comment]
@dotnet-bot please retest this

**Comment by tannergooding** [comment]
test windows_eta_open_prtest please

File: src/Compilers/VisualBasic/Portable/Parser/Parser.vb [hunk 1]
```diff
@@ -6078,18 +6078,26 @@ checkNullable:
...
+            If (opts.Features Is Nothing) OrElse (opts.Features.Count = 0) Then Return False
+            Dim value As String = Nothing
+            Return opts.Features.TryGetValue(feature.GetFeatureFlag, value)
+        End Function
+
+        Friend Shared Function CheckFeatureAvailability(opts As VisualBasicParseOptions, feature As Feature) As Boolean
             Dim required = feature.GetLanguageVersion()
-            Return CInt(required) <= CInt(languageVersion)
+            If CInt(required) <= CInt(opts.LanguageVersion) Then Return True
+            Return CheckFeatures(feature, opts)
```
Comment (AdamSpeight2008) [inline src/Compilers/VisualBasic/Portable/Parser/Parser.vb]: Simplification.
```vb
Return (CInt(requireed) <= CInt(opts.LanguageVersion)) OrElse CheckFeatures(feature, opts)
```

File: src/Compilers/VisualBasic/Portable/Parser/Parser.vb [hunk 2]
```diff
@@ -6078,18 +6078,25 @@ checkNullable:
...
         Friend Function CheckFeatureAvailability(feature As Feature) As Boolean
-            Return CheckFeatureAvailability(_scanner.Options.LanguageVersion, feature)
+            Return CheckFeatureAvailability(_scanner.Options, feature)
         End Function
 
-        Friend Shared Function CheckFeatureAvailability(languageVersion As LanguageVersion, feature As Feature) As Boolean
+        Shared Function CheckFeatures(feature As Feature, opts As VisualBasicParseOptions) As Boolean
+            If (opts.Features Is Nothing) OrElse (opts.Features.Count = 0) Then Return False
+            Dim value As String = Nothing
+            Return opts.Features.TryGetValue(feature.GetFeatureFlag, value)
```
Comment (AdamSpeight2008) [inline src/Compilers/VisualBasic/Portable/Parser/Parser.vb]: Should check to see if the `FeatureFlag Is Nothing` before checking features.

**Comment by AdamSpeight2008** [comment]
This change has revealed Issue #12234

**Comment by AdamSpeight2008** [comment]
@jaredpar  Ready For Review.

**Comment by gafter** [comment]
I have no idea why this change is proposed. There is no issue describing the improvement that it hopes to make.

**Comment by AdamSpeight2008** [comment]
@gafter The current check only check the language version of a feature flag, some may not have been approved for an official version. In particular if that feature is still in development or a prototype.
So we could pass use the command line an pass in that feature `feature:foo`, which is put in the `ParseOptions.Features` 
This PR checks `ParseOptions.Features` as well.

**Comment by gafter** [comment]
Do you have a test that illustrates the difference in behavior?

**Comment by AdamSpeight2008** [comment]
Not currently, but I'll add one.

**Comment by crummel** [comment]
Build triggered. sha1 is merged.

**Comment by AdamSpeight2008** [comment]
@gafter @jaredpar @AnthonyDGreen 
Would it be possible to get a review of this, as whether it gets accepted or not, affects the code in another pull request containing a VB language feature. (PR #12629)

File: src/Compilers/VisualBasic/Portable/Parser/Parser.vb [hunk 3]
```diff
@@ -6078,18 +6080,30 @@ checkNullable:
...
         End Function
 
-        Friend Shared Sub CheckFeatureAvailability(diagnostics As DiagnosticBag, location As Location, languageVersion As LanguageVersion, feature As Feature)
-            If Not CheckFeatureAvailability(languageVersion, feature) Then
-                Dim featureName = ErrorFactory.ErrorInfo(feature.GetResourceId())
-                diagnostics.Add(ERRID.ERR_LanguageVersion, location, languageVersion.GetErrorName(), featureName)
+        Friend Shared Function CheckFeatureAvailability(opts As VisualBasicParseOptions, feature As Feature) As Boolean
+            Dim required As LanguageVersion
+            Dim ok = feature.TryGetLanguageVersion(required)
+            Return ok AndAlso (CInt(required) <= CInt(opts.LanguageVersion)) OrElse CheckFeatures(feature, opts)
```
Comment (AdamSpeight2008) [inline src/Compilers/VisualBasic/Portable/Parser/Parser.vb]: Needs additional check `required <> LanguageVersion.None`, to force it to check features.

**Comment by gafter** [comment]
This PR introduces an unwanted public API change so no, it would not be accepted in this form.

**Comment by AdamSpeight2008** [comment]
What change?

File: src/Compilers/VisualBasic/Portable/Parser/Parser.vb (hunk 3 above)
Comment (AdamSpeight2008) [inline src/Compilers/VisualBasic/Portable/Parser/Parser.vb]: Correction: 
```
Return (ok AndAlso CInt(required) <= CInt(opts.LanguageVersion)) OrElse CheckFeatures(feature, opts)
```

**Comment by gafter** [comment]
> What change?

Did you notice that you changed `PublicAPI.Unshipped.txt`?

**Comment by jaredpar** [comment]
Didn't we discuss this on an issue thread where we said we would be against this type of change?

**Comment by gafter** [comment]
@jaredpar Sounds vaguely familiar, but I agree that this PR does not introduce functionality we intend to add to the compiler.

**Pull request closed without merging**